File: deepscore.py
import os
import scanpy as sc
import numpy as np
import tensorflow as tf
import pandas as pd
from sklearn.model_selection import train_test_split
from tensorflow import keras
from tensorflow.keras.constraints import max_norm
from tensorflow.keras.regularizers import l1_l2
import imageio


class PlotInference(keras.callbacks.Callback):
    
    def __init__(self, ds, adata, gif_name='mygif'):
        super(PlotInference, self).__init__()
        self.adata = adata
        self.filenames = []
        self.gif_name = gif_name
        self.ds = ds
        
    def on_epoch_end(self, epoch, logs=None):
        with tf.device('/CPU:0'):
            self.adata = self.ds.annotate(self.adata)
            filename = f'_ds_training{epoch}.png'
            sc.pl.umap(self.adata, color='Deepscore', legend_loc = 'on data',
                      save=filename, show=False)
            self.filenames.append(f'figures/umap{filename}')

    def on_train_end(self, logs=None):
        
        # build gif
        with imageio.get_writer(f'{self.gif_name}.gif', mode='I') as writer:
            for filename in self.filenames:
                print(filename)
                image = imageio.imread(filename)
                writer.append_data(image)

        # Remove files
        for filename in self.filenames:
            os.remove(filename)

            

class DeepScore():
    
    def __init__(self, hidden_nodes, n_features, n_labels, epochs=30, 
                 batch_size=32, activation="relu", dropout=True, 
                 dropout_rate=0.2, batchnorm=True, lr=0.001,
                 weight_reg=True, l1=0, l2=0):
        
        super(DeepScore, self).__init__()
        self.epochs = epochs
        self.batch_size = batch_size
        self.activation = activation
        self.dropout = dropout
        self.dropout_rate = dropout_rate
        self.lr = lr
        
        if weight_reg:
            self.w_norm = max_norm(2.)
        else:
            self.w_norm = None
    
        model = keras.Sequential(name="deepscore")
        model.add(keras.Input(shape=n_features,))
        if batchnorm:
            model.add(keras.layers.BatchNormalization())

        for i, n in enumerate(hidden_nodes):
            model.add(keras.layers.Dense(n, activation=activation, 
                    kernel_constraint=self.w_norm, name=f"dense{n}",
                    kernel_regularizer=l1_l2(l1=l1, l2=l2)))
            if dropout:
                model.add(keras.layers.Dropout(dropout_rate))
            if batchnorm:
                model.add(keras.layers.BatchNormalization())

        model.add(keras.layers.Dense(n_labels, activation='softmax', 
                                     name="output"))

        model.compile(
            optimizer=keras.optimizers.Adam(learning_rate=lr),
            loss=keras.losses.CategoricalCrossentropy(),
            metrics=[keras.metrics.CategoricalAccuracy()])

        model.summary()
        self.model = model
    

            
    def set_reference(self, adata, label_by, test_prop=0.25):
    
        # Turn each cell label into a hot one encoding
        labels = pd.get_dummies(adata.obs[label_by])
        self.label_dict = {i: k for i, k in enumerate(list(labels.columns))}

        xtrain, xtest, ytrain, ytest = train_test_split(adata.X, labels,
                                                        test_size = test_prop)

        xtrain = tf.constant(xtrain)
        xtest = tf.constant(xtest)
        ytrain = tf.constant(ytrain)
        ytest = tf.constant(ytest)

        self.splitted_data = xtrain, xtest, ytrain, ytest
    
    
    
    
    def train(self, val_split=0.1, earlystopping=True, lr_scheduler=False, 
              patience=2, create_gif=False, gif_dataset=None):
    
        xtrain, xtest, ytrain, ytest = self.splitted_data

        my_callbacks = [tf.keras.callbacks.TensorBoard(
            log_dir='./deepscore_logs')]
        
        if earlystopping:
            my_callbacks.append(
                tf.keras.callbacks.EarlyStopping(patience=patience))
        if lr_scheduler:
            my_callbacks.append(
                tf.keras.callbacks.LearningRateScheduler(lr_scheduler))
        if create_gif:
            if gif_dataset is None:
                print('Please specify a dataset in gif_dataset')
            else:
                my_callbacks.append(PlotInference(self, gif_dataset))

        self.model.fit(xtrain, ytrain, epochs=self.epochs, 
                       batch_size=self.batch_size, validation_split=val_split, 
                       verbose = True, callbacks=my_callbacks)

        print("\nEvaluating model performance on unseen data (test data):\n")
        evaluation = self.model.evaluate(xtest, ytest, batch_size=self.batch_size)
        print(f"\ntest loss: {round(evaluation[0], 5)}, test accuracy:' \
              '{round(evaluation[1], 5)}")
    
    
    
    def annotate(self, adata, pred_key='Deepscore'):

        tf.data.Dataset.from_tensor_slices(adata.X)
        test = self.model.predict(adata.X)

        indeces = np.argmax(test, axis=1)
        maxs = np.max(test, axis=1)
        predictions = []

        for m, i in zip(maxs, indeces):
            if m > 0.5:
                predictions.append(self.label_dict[i])
            else:
                predictions.append('Unclassified')

        adata.obs[pred_key] = predictions

        return(adata)        
        

    

File: R/deepscore.R


DeepScore <- function(hidden_nodes, common_features, n_labels, epochs=10, 
                 batch_size=32, activation="relu", dropout=TRUE, 
                 dropout_rate=0.2, batchnorm=TRUE, lr=0.001,
                 weight_reg=TRUE, l1=0, l2=0) {
        
    ds <- list(
        model = NULL,
        common_features = common_features,
        epochs = epochs,
        batch_size = batch_size,
        activation = activation,
        dropout = dropout,
        dropout_rate = dropout_rate,
        lr = lr,
        w_norm = NULL
        )
    
    n_features <- length(common_features)

    if (weight_reg) {
        ds$w_norm <- constraint_maxnorm(max_value = 2)
    }

    model <- keras_model_sequential()

    if (batchnorm) {
        model <- layer_batch_normalization(model)
    }
    
    for (i in seq_along(hidden_nodes)) {
        model <- model %>% 
                    layer_dense(hidden_nodes[i], activation=activation, input_shape = c(n_features),
                                kernel_constraint=ds$w_norm, name=glue("dense{hidden_nodes[i]}"),
                                kernel_regularizer=regularizer_l1_l2(l1=l1, l2=l2))
        if (dropout) {
            model <- layer_dropout(model, rate=dropout_rate)
        }
        if (batchnorm) {
            model <- layer_batch_normalization(model)
        }
        
        # N_features as input shape is only for the first layer
        n_features <- NULL
    }
    
    model <- model %>% 
                layer_dense(n_labels, activation="softmax", name="output")
    
    model %>% compile(
        optimizer = optimizer_adam(learning_rate=lr), 
        loss = 'categorical_crossentropy',
        metrics = c('accuracy')
    )
    
    #print(summary(model))
    
    ds$model <- model

    class(ds) <- append(class(ds), "DeepScore")
    return(ds)
}




set_reference <- function(ds, reference, labels, assay="RNA",
                          test_prop = 0.2) {

    reference <- reference[ds$common_features,]
    reference <- GetAssayData(reference, slot = "data", assay = assay)
    
    labels <- as.factor(unname(unlist(labels)))
    classes <- levels(labels)
    levels(labels) <- seq(1:length(levels(labels)))

    sizes <- unlist(lapply(levels(labels), function(x) length(labels[which(labels %in% x)])))
    names(sizes) <- levels(labels)

    coln <- colnames(reference)
    rown <- rownames(reference)
                           
    # Random sampling for training data but making sure we take from all clusters
    train.idx <- unlist(sapply(levels(labels), function(x) 
        sample(x = which(labels %in% x), 
        size = sizes[names(sizes) == x] * (1 - test_prop)))
        )
    train.data <- reference[, train.idx] %>% Matrix::t()
    train_y <- labels[train.idx]

    test.data <- reference[, -train.idx] %>% Matrix::t()
    test_y <- labels[-train.idx]
                       
    # Shuffling the training data
    model.train <- data.frame(train_y, train.data, check.names = F)
    model.train <- model.train[sample(seq(1:nrow(model.train))),]
    
    ds$split_data <- list(
        train_x = as.matrix(model.train[,-1]),
        train_y = to_categorical(model.train[,1]),
        test_x = test.data,
        test_y = to_categorical(test_y),
        classes = classes
        )
                
    return(ds)
}
                        
                        

train <- function(ds, plot = TRUE, earlystopping=TRUE, lr_scheduler=FALSE, 
              schedule=NULL, patience=2) {
    
    callbacks <- c(callback_tensorboard(log_dir = "./deepscore_logs"),
                   callback_progbar_logger())
    
    if (earlystopping) {
        callbacks <- c(callbacks, callback_early_stopping(
            monitor = "val_loss", 
            patience = patience, 
            restore_best_weights = TRUE
        ))
    }

    if (lr_scheduler) {
        callbacks <- c(callbacks, callback_learning_rate_scheduler(schedule))
    }
    
    history <- ds$model %>% fit(
        x = ds$split_data$train_x,
        y = ds$split_data$train_y,
        epochs = ds$epochs,
        batch_size = ds$batch_size,
        validation_split = .2,
        callbacks=callbacks,
        verbose = 2
    )
    
    test_performance <- ds$model %>% evaluate(ds$split_data$test_x,
                                               ds$split_data$test_y)
    print("Model evaluation on unseen data:")
    print(test_performance)
    
    if (plot) {
        print(plot(history))
    }
    
    return(ds)
}
                        
                        
                        
annotate <- function(ds, query, assay="RNA") {
    
    query.data <- GetAssayData(query, slot = "data", assay = assay)[ds$common_features,]
    query.data <- Matrix::t(query.data)

    probs <- ds$model %>% predict(query.data)
    probs <- data.frame(probs, check.names = F)
    names(probs) <- c("unclassified", ds$classes)
    predicted <- apply(probs, 1, function(x) {
        ifelse(max(x) > 0.5, 
               names(probs)[which(x==max(x))],
               "unclassified"
              )
    })

    query$ds_annotation <- factor(predicted)

    return(query)
}
